experimental/framework/fluid-static/src/serviceAudience.ts

```diff
@@ -12,67 +12,91 @@ import { IServiceAudience, IServiceAudienceEvents, IMember } from "./types";
 // Base class for providing audience information for sessions interacting with FluidContainer
 // This can be extended by different service-specific client packages to additional parameters to
 // the user and client details returned in IMember
-export class ServiceAudience
-  extends TypedEventEmitter<IServiceAudienceEvents<IMember>>
-  implements IServiceAudience<IMember> {
+export abstract class ServiceAudience<M extends IMember = IMember>
+  extends TypedEventEmitter<IServiceAudienceEvents<M>>
+  implements IServiceAudience<M> {
   protected readonly audience: IAudience;
 
+  /**
+   * Retain the most recent member list.  This is so we have more information about a member
+   * leaving the audience in the removeMember event.  It allows us to match the behavior of the
+   * addMember event where it only fires on a change to the members this class exposes (and would
+   * actually produce a change in what getMembers returns).  It also allows us to provide the
+   * client details in the event which makes it easier to find that client connection in a map
+   * keyed on the userId and not clientId.
+   * This map will always be up-to-date in a removeMember event because it is set once at
+   * construction and in every addMember event.
+   * It is mapped clientId to M to be better work with what the IAudience event provides
+   */
+  protected lastMembers: Map<string, M> = new Map();
+
   constructor(
       protected readonly container: Container,
   ) {
     super();
     this.audience = container.audience;
 
-    // Consolidating both the addition/removal of members
-    this.audience.on("addMember", () => {
-      this.emit("membersChanged", this.getMembers());
-    });
+    // getMembers will assign lastMembers so the removeMember event has what it needs
+    // in case it would fire before getMembers otherwise gets called the first time
+    this.getMembers();
 
-    this.audience.on("removeMember", () => {
-      this.emit("membersChanged", this.getMembers());
+    this.audience.on("addMember", (clientId: string, details: IClient) => {
+      if (this.shouldIncludeAsMember(details)) {
+        const member = this.getMember(clientId);
+        this.emit("addMember", clientId, member);
+        this.emit("membersChanged");
+      }
     });
 
-    this.container.on("connected", () => {
-      this.emit("membersChanged", this.getMembers());
+    this.audience.on("removeMember", (clientId: string) => {
+      if (this.lastMembers.has(clientId)) {
+        this.emit("removeMember", clientId, this.lastMembers.get(clientId));
+        this.emit("membersChanged");
+      }
     });
   }
 
+  protected abstract createServiceMember(audienceMember: IClient): M;
```

Comment:
I really like this conversion of `ServiceAudience` to be an abstract class where all each of the clients needed to do is fill in this one method to correctly parse out their user details!

Could we also add `createServiceMember` to the `IServiceAudience` interface?

Reply:
Are there scenarios where an external user would want to call `createServiceMember`?  Having it be abstract here means anyone extending `ServiceAudience` will need to implement it, but it didn't seem important to me as part of the public API

Reply:
Fair, it doesn't need to be accessed by any outsider calling it. Let's keep it private 😄